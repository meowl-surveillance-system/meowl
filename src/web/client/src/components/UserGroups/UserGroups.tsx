import React, { Component } from "react";
import { Container, Typography } from "@material-ui/core";
import List from "@material-ui/core/List";
import ListItem from "@material-ui/core/ListItem";
import ListSubheader from "@material-ui/core/ListSubheader";

import GroupForm from "../GroupForm/GroupForm";

interface Props { }
interface State {
  usernames: Array<string>;
  groups: Array<string>;
  groupsToUsersDict: Record<string, Array<string>>;
}

/**
 * A component for adding users to groups
 * TODO: Write tests for this component
 */
export default class UserGroups extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      usernames: [],
      groups: [],
      groupsToUsersDict: {},
    };
  }

  /**
   * Fetch the usernames and groups
   */
  componentDidMount() {
    fetch(`/auth/getUsernames`)
      .then((res) => res.json())
      .then((usernames) => this.setState({ usernames }))
      .catch((e) => console.error(e));
    fetch(`/api/getGroups`)
      .then((res) => res.json())
      .then((groups) => this.setState({ groups }))
      .then(() => {
        this.state.groups.map((groupId, index) => {
          return fetch(`/api/getGroupUsers/${groupId}`)
            .then((res) => res.json())
            .then((users) => this.setState({ groupsToUsersDict: {...this.state.groupsToUsersDict, [groupId]: users } }))
            .catch((e) => console.error(e))
        })
      })
      .catch((e) => console.error(e))
  }

  /**
   * Renders the list of usernames
   */
  renderUsernames = () => {
    return this.state.usernames.map((username, index) => {
      return ( 
        <ListItem
          key={index}
          button
        >
        {username}
        </ListItem>
      );
    });
  };

  /**
   * Render all groups
   */
  renderGroups = () => {
    return Object.keys(this.state.groupsToUsersDict).map((group, index) => {
      return ( 
        <List
          subheader={
            <ListSubheader color="inherit">
              <Typography variant="inherit">{group}</Typography>
            </ListSubheader>
          }
          key={index}
        >
        {this.renderGroupUsers(this.state.groupsToUsersDict[group])}
        </List>
      );
    });
  };

  /**
   * Render the usernames for the specified group
   * @params group - The specified group to be rendered
   */
  renderGroupUsers = (group: Array<string>) => {
    return group.map((username, index) => {
      return (
        <ListItem
          key={index}
          button
        >
          {username}
        </ListItem>
      );
    });
  }

  /**
   * Renders out the usernames and forms to submit users to groups
   */
  render() {
    return (
      <Container>
         <List
          subheader={
            <ListSubheader color="inherit">
              <Typography variant="inherit">Groups</Typography>
            </ListSubheader>
          }
        >
         {this.renderGroups()}
        </List>
         <List
          subheader={
            <ListSubheader color="inherit">
              <Typography variant="inherit">Usernames</Typography>
            </ListSubheader>
          }
        >
          {this.renderUsernames()}
        </List>
        <GroupForm />
      </Container>
    );
  }
}
