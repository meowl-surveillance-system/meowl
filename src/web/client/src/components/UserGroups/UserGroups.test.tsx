import React from "react";
import { render } from "@testing-library/react";
import UserGroups from "./UserGroups";

describe("UserGroup component", () => {

  it("should render the UserGroup component", () => {
    const { container } = render(
      <UserGroups />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});