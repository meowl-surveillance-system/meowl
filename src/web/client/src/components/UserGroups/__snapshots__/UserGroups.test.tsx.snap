// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserGroup component should render the UserGroup component 1`] = `
<div
  class="MuiContainer-root MuiContainer-maxWidthLg"
>
  <ul
    class="MuiList-root MuiList-padding MuiList-subheader"
  >
    <li
      class="MuiListSubheader-root MuiListSubheader-colorInherit MuiListSubheader-sticky MuiListSubheader-gutters"
    >
      <span
        class="MuiTypography-root"
      >
        Groups
      </span>
    </li>
  </ul>
  <ul
    class="MuiList-root MuiList-padding MuiList-subheader"
  >
    <li
      class="MuiListSubheader-root MuiListSubheader-colorInherit MuiListSubheader-sticky MuiListSubheader-gutters"
    >
      <span
        class="MuiTypography-root"
      >
        Usernames
      </span>
    </li>
  </ul>
  <main
    class="MuiContainer-root MuiContainer-maxWidthXs"
  >
    
    
    <h1
      class="MuiTypography-root MuiTypography-h5"
    >
      Add User to Group
    </h1>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Add user to a group. If groupId does not exist, a new group will be created with the provided groupId
    </p>
    <form
      novalidate=""
    >
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
          data-shrink="false"
        >
          GroupId
          <span
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            name="groupId"
            required=""
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-107 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-109"
            >
              <span>
                GroupId
                 *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
          data-shrink="false"
        >
          Username
          <span
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            name="username"
            required=""
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-107 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-109"
            >
              <span>
                Username
                 *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label"
        >
          Submit
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </form>
  </main>
</div>
`;
